lib/cyrusdb_twom.c

```diff
@@ -0,0 +1,2609 @@
+{
+    struct dcrock *drock = rock;
+    struct dbengine *db = NULL;
+    int r = myopen(drock->fname, drock->flags, &db, NULL);
+    if (r == CYRUSDB_NOTFOUND) {
+        syslog(LOG_INFO, "twom: no file to delayed checkpoint for %s",
+               drock->fname);
+    }
+    else if (r) {
+        syslog(LOG_ERR, "DBERROR: opening %s for checkpoint: %s",
+               drock->fname, cyrusdb_strerror(r));
+    }
+    else if (db->openfile->header.generation == drock->generation) {
+        // if it hasn't already happened
+        mycheckpoint(db);
+    }
+    else {
+        syslog(LOG_INFO, "twom: delayed checkpoint already done %s (%llu %llu)",
+               drock->fname, (LLU)db->openfile->header.generation, (LLU)drock->generation);
+    }
+    if (db) myclose(db);
+}
+
+static int commit_locked(struct dbengine *db, struct txn *txn)
+{
+    if (txn != db->write_txn) {
+        struct txn **ptr;
+        for (ptr = &db->read_txn; *ptr; ptr = &(*ptr)->next)
+            if (*ptr == txn) break;
+        assert(*ptr);
+        _remove_txn(ptr);
+        tm_cleanup(db);
+        return 0;
+    }
+
+    int r = 0;
+
+    assert(db);
+    assert(txn);
+
+    struct tm_loc *loc = &db->loc;
+    struct tm_file *file = loc->file;
+    assert(file == txn->file);
+    assert(file->has_lock == 2);
+
+    struct tm_header *header = &file->header;
+
+    /* no need to commit if we're not dirty */
+    if (!(header->flags & DIRTY))
+        goto done;
+
+    size_t headlen = 16;
+    size_t reclen = 24;
+
+    // could re-map, but it WON'T change the header
+    r = tm_ensure(db, txn->end + reclen);
+    if (r) goto done;
+
+    char *base = file->base + txn->end;
+    memset(base, 0, reclen); // zero out the whole thing before we set just the bits we want
+
+    *((uint8_t *)(base)) = COMMIT;
+    *((uint64_t *)(base+8)) = htonll(header->current_size);
+    *((uint32_t *)(base+16)) = htonl(file->csum(base, headlen));
+
+    txn->end += reclen;
+    loc->end = txn->end;
+    file->dirty = 1;
+
+    /* commit ALL outstanding changes first, before
+     * rewriting the header */
+    r = tm_commit(db, txn->end);
+    if (r) goto done;
+
+    /* finally, update the header and commit again */
+    header->num_commits++;
+    header->current_size = txn->end;
+    header->flags &= ~DIRTY;
+    r = commit_header(db, header);
+    if (r) goto done;
+
+    if (!db->nocompact
+           && header->dirty_size > MINREWRITE
+           && header->current_size < 4 * header->dirty_size) {
+        // delay the checkpoint until the user isn't waiting
+        struct dcrock *drock = xzmalloc(sizeof(struct dcrock));
+        drock->fname = xstrdup(db->fname);
+        drock->flags = 0;
+        libcyrus_delayed_action(drock->fname, _delayed_checkpoint,
+                                _delayed_checkpoint_free, drock);
+    }
+
+ done:
+    if (r) {
+        int r2;
+
+        /* error during commit; we must abort */
+        r2 = abort_locked(db, txn);
```

Comment:
```c
int r2;
r2 = abort_locked(db, txn);
```
can be put on one line.  But since the value of `r2` is used only once, that variable can be completely removed:
```c
/* error during commit; we must abort */
if (abort_locked(db, txn)) {
           xsyslog(LOG_ERR, "DBERROR: commit AND abort failed", "filename=<%s>", db->fname);
}
return r;

```

Reply:
Yes, but that makes it hard to instrument tests (or step though with a debugger).  Variables cost nothing when the compiler optimises them out.